ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,437 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
+    DeviceString.MaximumLength = DeviceString.Length;
+
+    DPRINT("Device string is '%wZ'\n", &DeviceString);
+
+    if (MungedString != NULL)
+    {
+        /* Allocate a munged path string */
+        MungedStringReal.Length = 0;
+        MungedStringReal.MaximumLength = MungedPrefix.Length + DeviceString.Length + sizeof(UNICODE_NULL);
+        MungedStringReal.Buffer = ExAllocatePoolWithTag(PagedPool, MungedStringReal.MaximumLength, TAG_IO);
+        if (!MungedStringReal.Buffer)
+        {
+            DPRINT1("ExAllocatePoolWithTag() failed\n");
+            return STATUS_INSUFFICIENT_RESOURCES;
+        }
+
+        /* Fill it by munged prefix and our previously created device string */
+        Status = RtlAppendUnicodeStringToString(&MungedStringReal, &MungedPrefix);
+        NT_ASSERT(NT_SUCCESS(Status));
+
+        Status = RtlAppendUnicodeStringToString(&MungedStringReal, &DeviceString);
+        NT_ASSERT(NT_SUCCESS(Status));
+
+        DPRINT("Munged string is '%wZ'\n", &MungedStringReal);
+    }
+
+    DPRINT("GUID string is '%wZ'\n", &GuidStringReal);
+    DPRINT("Reference string is '%wZ'\n", &ReferenceStringReal);
+
+    /* Store received parts if the parameters are not null */
+    if (PrefixString != NULL)
+    {
+        if (RtlPrefixUnicodeString(&KernelModePrefix, SymbolicLinkName, FALSE))
+        {
+            PrefixString->Buffer = SymbolicLinkName->Buffer;
+            PrefixString->Length = KernelModePrefix.Length;
+            PrefixString->MaximumLength = PrefixString->Length;
+        }
+        else if (RtlPrefixUnicodeString(&UserModePrefix, SymbolicLinkName, FALSE))
+        {
+            PrefixString->Buffer = SymbolicLinkName->Buffer;
+            PrefixString->Length = UserModePrefix.Length;
+            PrefixString->MaximumLength = PrefixString->Length;
+        }
+
+        DPRINT("Prefix string is '%wZ'\n", PrefixString);
+    }
+
+    if (MungedString != NULL)
+        *MungedString = MungedStringReal;
     else
-        InstanceKeyRealP = &InstanceKeyReal;
+        RtlFreeUnicodeString(&MungedStringReal);
 
-    *GuidKeyRealP = NULL;
-    *DeviceKeyRealP = NULL;
-    *InstanceKeyRealP = NULL;
+    if (GuidString != NULL)
+        *GuidString = GuidStringReal;
 
-    RtlInitUnicodeString(&BaseKeyU, BaseKeyString);
+    if (ReferenceString != NULL)
+        *ReferenceString = ReferenceStringReal;
 
-    /* Open the DeviceClasses key */
-    InitializeObjectAttributes(&ObjectAttributes,
-                               &BaseKeyU,
-                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
-                               NULL,
-                               NULL);
-    Status = ZwOpenKey(&ClassesKey,
-                       DesiredAccess | KEY_ENUMERATE_SUB_KEYS,
-                       &ObjectAttributes);
+    if (ReferenceStringPresent != NULL)
+        *ReferenceStringPresent = ReferenceStringReal.Length != sizeof(WCHAR);
```

Comment:
So it would be TRUE even if `ReferenceStringReal.Length` is == 0 ?
Would this need to be something like `ReferenceStringReal.Length > sizeof(WCHAR);` instead?

Reply:
Indeed, you're right. :sweat_smile: It always will be `TRUE` on zero length too. So your suggestion is better, thanks.

Reply:
Fixed, thank you again! :+1: